### PR TITLE
Use `fingerprint` file identity by default and migrate file state from `native` or `path`

### PR DESCRIPTION
**Pull request by belimawr**
## Proposed commit message

This commit changes the default `file_identity` from `native` to
`fingerprint`, any previous state from `native` (or `path`) is
automatically migrated to `fingerprint` whe Filestream is starting.

The Filestream input has always had the [ability to update file identifiers](https://github.com/elastic/beats/blob/4278366ab03221e8b62183dc06f9505f6ccc5209/filebeat/input/filestream/prospector.go#L104-L122),
however it never worked as expected, leading to full data duplication
when changing the file identity. This commit fixes it to allow
changing the file identity from `native` (inode + device ID) and
`path` to `fingerprint` without any data duplication.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact
Because the `fingerprint` is the new default file identity, files are now only ingested when they reach at least 1024 bytes. The old default behaviour can be enabled by setting the file identity to `native` and disabling the fingerprint in the scanner.
```yaml
filebeat.inputs:
  - type: filestream
    id: "8.x-default-behaviour"
    paths:
      - /tmp/flog.log
    file_identity.native: ~
    prospector:
      scanner:
        fingerprint.enabled: false
```

## Author's Checklist
 - [x] Test with dynamic config reload
 - [x] Test with Kubernetes
 - [x] Test with Elastic-Agent
 - [x] Fix all the tests that break with the new behaviour
 - [x] Investigate which integration tests are going to break in the Elastic-Agent repo

Regarding the Elastic-Agent integration tests, most tests actually use the `log` input because when they were written, Filestream was not available as an integration package. The very few other test that use Filestrem either generate a log file large enough or are skipped as flaky.

## How to test this PR locally

1. Create a log file with at least a few log lines and more than 1kb (e.g: `/tmp/flog.log`, 15 log lines), you can use [`flog`](https://github.com/mingrammer/flog) with Docker:
    ```
    docker run -it --rm mingrammer/flog -n 15 > /tmp/flog.log
    ```
3. Start Filebeat with the following configuration

    <details><summary>filebeat.yml (native)</summary>
    <p>

    ```yaml
    filebeat.inputs:
      - type: filestream
        id: "test-migrate-ID"
        paths:
          - /tmp/flog.log
        file_identity.native: ~
        prospector:
          scanner:
            check_interval: 0.1s
            fingerprint.enabled: false
    
    queue.mem:
      flush.timeout: 0s
    
    output.file:
      path: ${path.home}
      filename: "output-file"
      rotate_on_startup: false
    
    logging:
      level: debug
      selectors:
        - input
        - input.filestream
        - input.filestream.prospector
      metrics:
        enabled: false
    ```

    </p>
    </details> 

4. Wait until the file is fully ingested (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
5. Ensure all events have been published to the output (`wc -l ./output-file*` should return 15)
6. Stop Filebeat
7. Change the [file identity](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-filestream.html#_file_identity_2) to `fingerprint`. It's the new default, hence it's not explicitly set.
    <details><summary>filebeat.yml (fingerprint)</summary>
    <p>

    ```yaml
    filebeat.inputs:
      - type: filestream
        id: "test-migrate-ID"
        paths:
          - /tmp/flog.log
        prospector:
          scanner:
            check_interval: 0.1s

    queue.mem:
      flush.timeout: 0s

    output.file:
      path: ${path.home}
      filename: "output-file"
      rotate_on_startup: false

    logging:
      level: debug
      selectors:
        - input
        - input.filestream
        - input.filestream.prospector
      metrics:
        enabled: false
    ```

    </p>
    </details> 
8. Start Filebeat
9. Wait until the Filebeat "finds the end of the file" (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
10. Ensure no extra event was published ((`wc -l ./output-file*` should still return 15)
11. Add 10 more lines to the file:
    ```
    docker run -it --rm mingrammer/flog -n 10 >> /tmp/flog.log
    ```
13. Wait until the new lines are ingested (wait for `End of file reached: /tmp/flog.log; Backoff now.` in the logs)
14. Ensure all events have been published to the output with no duplication (`wc -l ./output-file*` should return 25)

## Related issues

- Closes https://github.com/elastic/beats/issues/40197

## Use cases
Dealing with identity reuse (e.g: inode reuse) without facing re-ingestion of data with Filestream input

~~## Screenshots~~

### CODE REVIEW

**Comment by rdner** [comment]
Let's make sure it's also tested with dynamic config reload and with the Elastic Agent control protocol.

When I worked on `take_over` (`log`->`filestream` input migration) I discovered that we have separate code paths for applying dynamic configuration and it requires special handling for state changes.

I'm not saying it's not handled here, just we need to include this into testing procedures.

**Comment by belimawr** [comment]
> Let's make sure it's also tested with dynamic config reload and with the Elastic Agent control protocol.
> 
> When I worked on `take_over` (`log`->`filestream` input migration) I discovered that we have separate code paths for applying dynamic configuration and it requires special handling for state changes.
> 
> I'm not saying it's not handled here, just we need to include this into testing procedures.

Thanks Denis! Do you mean at least a manual test or an integration test?

**Comment by belimawr** [comment]
> Let's make sure it's also tested with dynamic config reload and with the Elastic Agent control protocol.
> 
> When I worked on `take_over` (`log`->`filestream` input migration) I discovered that we have separate code paths for applying dynamic configuration and it requires special handling for state changes.
> 
> I'm not saying it's not handled here, just we need to include this into testing procedures.

Thanks Denis! Do you mean at least a manual test or an integration test?

The prospector initialisation happens much after any code path for starting/configuring an input, it should be totally agnostic from how the input was configured started. So I believe those cases are also covered. However, I do agree it is good to at least perform some manual test, just to be on the safe side.

File: filebeat/docs/inputs/input-filestream-file-options.asciidoc [hunk 1]
```diff
@@ -542,50 +542,50 @@ indirectly set higher priorities on certain inputs by assigning a higher
...
+The states of files generated by `native` file identity can be migrated to `fingerprint`.
 
 [source,yaml]
 ----
 file_identity.native: ~
 ----
 
 *`path`*:: To identify files based on their paths use this strategy.
-
++
```
Comment (belimawr) [inline filebeat/docs/inputs/input-filestream-file-options.asciidoc:572]: This `+` ensures the block below is on the same indentation as the title.

**Comment by belimawr** [comment]
The Windows test failure is unrelated to this PR, I created a flaky test issue: https://github.com/elastic/beats/issues/42059

**Comment by cmacknz** [comment]
> files are now only ingested when they reach at least 1024 bytes.

This will be breaking for somebody, let's only keep the change of default identity in 9.0.

It would likely be helpful to backport the code here but keep the default identity in 8.x unchanged to make future backports easier.

**Comment by jlind23** [comment]
> I merged main onto this branch/PR, let's see if CI gets green with a re-run

Looks like only the linter is unhappy.

**Comment by belimawr** [comment]
> > I merged main onto this branch/PR, let's see if CI gets green with a re-run
> 
> Looks like only the linter is unhappy.

I touched too many files, all those warnings are from bits of code I didn't touch. Because this PR is rather large, I wasn't planning on fixing all those lint warnings to reduce the changes that need reviewing.

However, if you insist, I can fix them.

**Comment by belimawr** [comment]
> > files are now only ingested when they reach at least 1024 bytes.
> 
> This will be breaking for somebody, let's only keep the change of default identity in 9.0.
> 
> It would likely be helpful to backport the code here but keep the default identity in 8.x unchanged to make future backports easier.

I've been thinking about the best way to do this:
- Get this PR merged
- Let mergify create the backport
- Update the backport PRs by reverting the commits that change the defaults and docs
- Ask someone to review/test just to be on the safe side as we don't want any breaking change to slip into 8.x branch
- Merge into 8.x once approved and CI green

File: filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl [hunk 2]
```diff
@@ -439,7 +439,7 @@ filebeat.inputs:
 
   # Method to determine if two files are the same or not. By default
   # the Beat considers two files the same if their inode and device id are the same.
```
Comment (andrzej-stencel) [inline filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl:441]: Is this comment still true if the default method is now `fingerprint`?

File: filebeat/input/filestream/prospector_test.go [hunk 3]
```diff
@@ -97,13 +101,13 @@ func TestProspector_InitCleanIfRemoved(t *testing.T) {
 }
 
 func TestProspector_InitUpdateIdentifiers(t *testing.T) {
-	f, err := ioutil.TempFile("", "existing_file")
+	f, err := os.CreateTemp("", "existing_file")
```
Comment (leehinman) [inline filebeat/input/filestream/prospector_test.go:104]: nit.  I think using t.TempDir() would be preferred.  That way we know the temp file will get deleted when the test exits.

**Review by leehinman — APPROVED** [review]

File: filebeat/input/filestream/prospector_test.go (hunk 3 above)
Comment (belimawr) [reply filebeat/input/filestream/prospector_test.go:104]: You're right, that's one of the lint issues I fixed, so I just replaced the `ioutil` call by its equivalent.

I'll make the change you suggested.

File: filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl (hunk 2 above)
Comment (belimawr) [reply filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl:441]: Thanks!

**Comment by belimawr** [comment]
I've enabled auto merge, it should get merged after CI runs :D

**Review by leehinman — APPROVED** [review]

File: filebeat/input/filestream/prospector.go [hunk 4]
```diff
@@ -113,12 +143,58 @@ func (p *fileProspector) Init(
...
-		return "", fm
+		previousIdentifierKey := newID(oldIdentifier.GetSource(
+			loginp.FSEvent{
+				NewPath:    fm.Source,
+				Descriptor: fd,
+			}))
+
+		// If the registry key and the key generated by the old identifier
+		// do not match, log it at debug level and do nothing.
+		if previousIdentifierKey != registryKey {
```
Comment (mauri870) [inline filebeat/input/filestream/prospector.go:186]: According to the comment it is missing a debug log here.

**Pull request merged**